### PR TITLE
Plotting for ZXDiagrams and ZXGraphs, Equality for ZXDiagrams and converting ZX to ZXW

### LINKED ISSUE #100: Cannot Plot ZXDiagram with YaoPlots
Dear Developers.

Please help me to get this library to Plot ZXDiagrams.
I cannot run the examples provided in this repository.

After installing ZXCalculus and YaoPlots. I tried running the examples from the 
shell, repl and in Pluto.
```bash
julia example/ex1.jl 
julia example/ex2.jl
```
 
Everytime I get the a *. See Stacktrace:

```
LoadError: BoundsError: attempt to access 1-element Vector{Int64} at index [2]
Stacktrace:
 [1] getindex
   @ ./array.jl:805 [inlined]
 [2] indexed_iterate
   @ ./tuple.jl:87 [inlined]
 [3] layout2locs(zxd::ZXDiagram{Int64, ZXCalculus.Phase})
   @ YaoPlots ~/.julia/packages/YaoPlots/U4hOY/src/zx_plot.jl:108
 [4] plot(zxd::ZXDiagram{Int64, ZXCalculus.Phase}; size_x::Nothing, size_y::Nothing, kwargs::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ YaoPlots ~/.julia/packages/YaoPlots/U4hOY/src/zx_plot.jl:176
 [5] plot(zxd::ZXDiagram{Int64, ZXCalculus.Phase})
   @ YaoPlots ~/.julia/packages/YaoPlots/U4hOY/src/zx_plot.jl:171
 [6] top-level scope
   @ ~/src/quantum-circuits/software/ZXCalculus.jl/example/ex2.jl:84
```

PS: Should this issue be moved to YaoPlots?

### PR DESCRIPTION
**Pull request by contra-bit**
Dear Maintainers,

This pull request introduces several improvements, including the addition of optional plotting functionality for  ZXDiagrams and ZXGraphs and closes #101 and fixes #100. The plotting function  uses Vega and DataFrames, which will only be loaded when necessary to improve load times and reduce memory usage. The color scheme of spider graphs has been updated to be more accessible for individuals with Color Vision Deficiency (CVD) and improved label readability. Further information on accessibility can be found at <[https://zxcalculus.com>](https://zxcalculus.com).

Additional testing has been implemented for both ZXDiagrams and ZXGraphs, with plots added to most tests and the creation of a new notebook focused on equivalence checking. Existing tutorial notebooks have also been refined, and users will now benefit from automated diagram generation within these notebooks.

Furthermore, the YaoHIR package has gained the capability to import OpenQASM files, enabling easy conversion from openQASM to ZXDiagrams and providing another useful approach to building. optimizing and verifying quantum circuits. 

The pull request further includes various bug fixes and improvements, such as addressing issues with the match and apply function for the ZXGraph :id rule, enhancing the overall user experience for plots, resolving testing issues, and handling miscellaneous bugs.

Regards and thanks for reviewing my pull request

### CODE REVIEW

**Comment by exAClior** [comment]
Hello @contra-bit , thank you very much for working on adding `plot` into `ZXCalculus.jl`.  I was trying out the notebook `tutorial.jl`. I was able to see the new plots, they looked cool! But were you able to install all the required packages that are at the beginning of this notebook? I could not do it. I noticed that you added `CompilerPluginTools` to dependency. That was causing a little trouble for us. It will probably cause more trouble for us when `Julia v1.10` is out. Roger is planning on updating YaoHIR and removing this completely see [link](https://github.com/JuliaCompilerPlugins/CompilerPluginTools.jl/pull/10#issuecomment-1619228573). Is this a hard requirement for your future work regarding this dependency?

**Comment by contra-bit** [comment]
Good catch. It is not required.

I only added `CompilerPluginTools` as an explicit dependency because it was causing trouble for me, when it was being  imported as a dependency by YaoHIR. 
The import failed because the version required could not be resolved.
The only way for me to resolve the `YaoHIR` dependency was to add the master branch from the git repo because the julia package had the wrong version. 
In order for it to be imported correctly I ran the following command:

```julia
pkg> add https://github.com/JuliaCompilerPlugins/CompilerPluginTools.jl
```
Then I added the equivalent command to the notebook as well. 

```julia
pkg> add https://github.com/JuliaCompilerPlugins/CompilerPluginTools.jl
```

**Comment by ChenZhao44** [comment]
@contra-bit Thanks for the PR. I suggest removing the dependency that is only used in the notebook, and keeping the main package as clean as possible. Perhaps it would be better to set up a separate environment in the `notebook` folder to include `CompilerPluginTools` and `DataFrames`.

**Comment by contra-bit** [comment]
I am thrilled to share several improvements in this pull request! I have introduced plotting functionality as an optional extension using Vega and DataFrames. This change ensures that these dependencies will only be loaded when needed, leading to faster load times and reduced memory usage.

To enhance accessibility, I have updated the color scheme of spider graphs so that they are distinguishable for people with Color Vision Deficiency (CVD), and improved the readability of phase labels. For further details, kindly consult the accessibility guidelines provided at zxcalculus.com.

My latest contributions enhance testing for both ZXDiagrams and ZXGraphs by adding plots to most tests, along with creating a brand-new notebook dedicated to equivalence checking and refining the existing tutorial notebook. Users will benefit from automatic generation of diagrams within these notebooks.

Moreover, YaoHIR has gained the ability to import OpenQASM files, allowing seamless conversion to ZXDiagrams—providing yet another practical method to construct quantum circuits. I will try to get all my contributions merged and then update the Pkg in the notebooks.

As part of my efforts, I fixed the match and apply function for the  ZXGraph :id rule while adding the new equivalence checking functionality. Now, it is possible to determine if two diagrams are identical through the built-in equivalency checks.

Last but not least, I took care of numerous housekeeping tasks such as updating imports, resolving problems discovered during testing, enhancing overall user experience for plots, and dealing with miscellaneous glitches

**Comment by ChenZhao44** [comment]
@contra-bit Thanks for working on this PR. It looks decent. Before merging this PR, I have only two minor suggestions.
- You don't need to keep the original `tutorial.jl` notebook since it is out-of-date.
- There are a lot of indent changes in this PR. The default indent of this repo is 4, but some of the indents are set to 2. These changes have created a lot of unnecessary differences, making it difficult to keep track of the real changes. Could please you fix the indent issue?

**Comment by Roger-luo** [comment]
FYI you don't have to use the package env for notebooks, every Pluto notebook has its own env by default.

**Comment by contra-bit** [comment]
I hope I have fixed the indentation issues, further I have increased the test coverage and added a .JuliaFormatter.toml that species that the default indentation is 4.

The notebooks are now configured to use the Yao upstream repositories. 
My next goal is to improve the notebooks and improve the documentation.

File: notebooks/zx_intro.jl [hunk 1]
```diff
@@ -11,8 +11,8 @@ begin
     Pkg.add("OpenQASM")
     Pkg.add("Vega")
     Pkg.add("DataFrames")
-    Pkg.add("YaoHIR.jl")
-    Pkg.add("ZXCalculus.jl")
+    Pkg.add("YaoHIR")
+    Pkg.add("ZXCalculus")
     Pkg.add(
         url = "https://gitlab.informatik.uni-bremen.de/qce/quantumequivalencechecker.jl",
         rev = "feat/zx",
```
Comment (Roger-luo) [inline notebooks/zx_intro.jl:18]: can we use Pluto builtin env instead of doing this?

File: src/ZX/rules.jl [hunk 2]
```diff
@@ -896,6 +909,11 @@ function check_rule(::Rule{:id}, zxg::ZXGraph{T, P}, vs::Vector{T}) where {T, P}
                 if spider_type(zxg, v1) == SpiderType.Z && spider_type(zxg, v3) == SpiderType.Z
                     return true
                 end
+                if ((spider_type(zxg, v1) == SpiderType.In || spider_type(zxg, v1) == SpiderType.Out) && 
+                    (spider_type(zxg, v3) == SpiderType.In || spider_type(zxg, v3) == SpiderType.Out))
+                  return true
+                end
+
             else phase(zxg, v2) == 1
```
Comment (ChenZhao44) [inline src/ZX/rules.jl:917]: should also be `elseif phase(zxg, v2) == 1`

File: src/ZX/rules.jl [hunk 3]
```diff
@@ -912,19 +930,26 @@ function rewrite!(::Rule{:id}, zxg::ZXGraph{T, P}, vs::Vector{T}) where {T, P}
         set_phase!(zxg, v2, zero(P))
         set_phase!(zxg, v1, -phase(zxg, v1))
         zxg.phase_ids[v1] = (zxg.phase_ids[v1][1], -zxg.phase_ids[v1][2])
+      end
+      if ((spider_type(zxg, v1) == SpiderType.In || spider_type(zxg, v1) == SpiderType.Out) && 
+          (spider_type(zxg, v3) == SpiderType.In || spider_type(zxg, v3) == SpiderType.Out))
+          rem_spider!(zxg, v2)
+          add_edge!(zxg, v1, v3, EdgeType.SIM)
```
Comment (ChenZhao44) [inline src/ZX/rules.jl:937]: Do you need to check the edge type for the edge (v1, v2) and (v2, v3)? For example, if only one of them are Hadamard edge, then the resulting edge should also be a Hadamard edge.

File: src/ZXW/utils.jl [hunk 4]
```diff
@@ -667,3 +667,14 @@ function substitute_variables!(
     end
     return zxwd
 end
+
+
+function convert_to_zxwd(root::YaoHIR.BlockIR)
+    # FIXME add test
```
Comment (ChenZhao44) [inline src/ZXW/utils.jl:673]: add tests?

File: test/ancilla_extraction.jl [hunk 5]
```diff
@@ -17,14 +17,14 @@ zxd = gen_phase_gadget()
 zxg = full_reduction(zxd)
 anc_circ = ancilla_extraction(zxg)
 
-# plot(anc_circ)
+@test plot(anc_circ) !== nothing
```
Comment (ChenZhao44) [inline test/ancilla_extraction.jl:20]: better to use `isnothing`

File: test/ancilla_extraction.jl [hunk 6]
```diff
@@ -17,14 +17,14 @@ zxd = gen_phase_gadget()
...
 zxd_swap = ZXDiagram(2)
 pushfirst_gate!(zxd_swap, Val(:SWAP), [1, 2])
-# plot(zxd_swap)
+plot(zxd_swap)
 convert_to_chain(zxd_swap)
 
 zxg_swap = ZXGraph(zxd_swap)
 zxd_anc = ancilla_extraction(zxg_swap)
-# plot(zxd_anc)
+@test plot(zxd_anc)  !== nothing
```
Comment (ChenZhao44) [inline test/ancilla_extraction.jl:29]: use `isnothing`

File: test/challenge.jl [hunk 7]
```diff
@@ -208,4 +208,5 @@ for i = 1:5
     push!(zxg.outputs, i+23)
 end
 
+@test plot(zxg) !== nothing
```
Comment (ChenZhao44) [inline test/challenge.jl:211]: use `isnothing`

File: test/circuit_extraction.jl [hunk 8]
```diff
@@ -38,3 +38,9 @@ zxg2 = clifford_simplification(zxd)
 
 zxg3 = full_reduction(zxd)
 cir = circuit_extraction(zxg3)
+
+# Plotting
+@test plot(zxd) !== nothing
```
Comment (ChenZhao44) [inline test/circuit_extraction.jl:43]: use `isnothing`

File: test/phase_teleportation.jl [hunk 9]
```diff
@@ -75,6 +75,8 @@ function gen_cir()
 end
 
 cir = gen_cir()
+@test plot(cir)!== nothing
```
Comment (ChenZhao44) [inline test/phase_teleportation.jl:78]: use `isnothing`

File: test/plots.jl [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+
+
+# Othertests for ZXGraphs and ZXDigram are embededd into the zx_graph and zx_diagram testsets
+zxd = ZXDiagram(3)
+ZX.insert_spider!(zxd, 1, 2, SpiderType.H)
+ZX.insert_spider!(zxd, 1, 2, SpiderType.X)
+ZX.insert_spider!(zxd, 1, 2, SpiderType.Z)
+zxg = ZXGraph(zxd)
+
+@test plot(zxd) !== nothing
```
Comment (ChenZhao44) [inline test/plots.jl:12]: use `isnothing`

File: test/rules.jl [hunk 11]
```diff
@@ -10,6 +10,7 @@ matches = match(Rule{:f}(), zxd)
 rewrite!(Rule{:f}(), zxd, matches)
 @test sort!(spiders(zxd)) == [1, 3]
 @test phase(zxd, 1) == phase(zxd, 3) == 3 // 4
+@test plot(zxd) !== nothing
```
Comment (ChenZhao44) [inline test/rules.jl:13]: use `isnothing`

File: test/zx_diagram.jl [hunk 12]
```diff
@@ -7,11 +7,13 @@ v_t = [SpiderType.X, SpiderType.Z, SpiderType.X]
 zxd = ZXDiagram(g, v_t, ps)
 zxd2 = ZXDiagram(g, Dict(zip(1:3, v_t)), Dict(zip(1:3, ps)))
 @test zxd.mg == zxd2.mg && zxd.st == zxd2.st && zxd.ps == zxd2.ps
+@test plot(zxd) !== nothing
```
Comment (ChenZhao44) [inline test/zx_diagram.jl:10]: use `isnothing`

File: src/ZXW/utils.jl (hunk 4 above)
Comment (contra-bit) [reply src/ZXW/utils.jl:673]: Test is in ir.jl in line 57. Forgot to remove the FIXME

File: notebooks/zx_intro.jl (hunk 1 above)
Comment (contra-bit) [reply notebooks/zx_intro.jl]: Sure

File: src/ZX/rules.jl (hunk 3 above)
Comment (contra-bit) [reply src/ZX/rules.jl:937]: Your assumption seems to be correct. I replaced the and with and or

**Comment by contra-bit** [comment]
I have addressed the feedback and think this can be merged now. @ChenZhao44 

Thank you for being patient.

File: src/ZX/circuit_extraction.jl [hunk 13]
```diff
@@ -285,7 +285,7 @@ function update_frontier!(zxg::ZXGraph{T, P}, gads::Set{T}, frontier::Vector{T},
         v = frontier[i]
         nb_v = neighbors(zxg, v)
         u = findfirst([u in gads for u in nb_v])
-        if u !== nothing
+          if !isnothing(u)
```
Comment (exAClior) [inline src/ZX/circuit_extraction.jl:288]: indentation is still messup, could you please check your formatting script? did it include this directory?

File: src/ZX/equality.jl [hunk 14]
```diff
@@ -0,0 +1,19 @@
+"""
+    verify_equality(zxd_1::ZXDiagram, zxd_2::ZXDiagram)
+
+checks the equivalence of two different ZXDiagrams
+
+"""
+function verify_equality(zxd_1::ZXDiagram, zxd_2::ZXDiagram)
+  merged_diagram = concat!(zxd_1, dagger(zxd_2))
```
Comment (exAClior) [inline src/ZX/equality.jl:8]: same thing here, the indentation is messed up

File: src/ZX/zx_diagram.jl [hunk 15]
```diff
@@ -629,3 +638,180 @@ function safe_convert(::Type{T}, x::Real) where T<:Rational
...
+    end
+    return -1
+end
+
+"""
+    import_non_in_out!(d1::ZXDiagram{T,P}, d2::ZXDiagram{T,P}, v2tov1::Dict{T,T}) where {T,P}
+
+Add non input and output spiders of d2 to d1, modify d1. Record the mapping of vertex indices.
+"""
+function import_non_in_out!(
```
Comment (exAClior) [inline src/ZX/zx_diagram.jl:677]: indentation problem here

File: src/ZX/zx_graph.jl [hunk 16]
```diff
@@ -299,7 +302,7 @@ function spider_sequence(zxg::ZXGraph{T, P}) where {T, P}
             spider_seq[q] = Vector{T}()
         end
         for v in vs
-            if qubit_loc(zxg, v) !== nothing
+          if !isnothing(qubit_loc(zxg, v))
```
Comment (exAClior) [inline src/ZX/zx_graph.jl:305]: indentation here

File: test/runtests.jl [hunk 17]
```diff
@@ -1,4 +1,15 @@
 using ZXCalculus, Documenter, Test
+using Vega
+using DataFrames
+
+
+@testset "plots.jl" begin
+  include("plots.jl")
```
Comment (exAClior) [inline test/runtests.jl:7]: indentation here

**Review by exAClior — COMMENTED** [review]
Hello @contra-bit , sorry for jumpping in. I rechecked your commits based on @ChenZhao44  and @Roger-luo 's comments. There are still a few outstanding problems.

File: .github/workflows/Documentation.yml
[the remark this answers is not in the document]
Comment (contra-bit) [reply .github/workflows/Documentation.yml:1]: I missed that it was in CI file. I have removed it

**Comment by contra-bit** [comment]
I reformatted the mentioned files by hand. Thank you for the thorough review.

**Comment by exAClior** [comment]
Great! This commit looks good to me. Let's see if Dr. Zhao and Dr. Luo has any further comments!

**Review by ChenZhao44 — APPROVED** [review]

**Comment by Roger-luo** [comment]
LGTM

**Pull request merged**